From 63709c4afa714c39bad0a0697c15f45e23c16c6f Mon Sep 17 00:00:00 2001
From: Lucas Hoffmann <lucc@posteo.de>
Date: Fri, 1 Oct 2021 22:07:30 +0200
Subject: [PATCH] Use :Man! instead of :set ft=man

This is needed for compatibility with neovim >= 0.5.1.  The upstream
note is at
https://github.com/neovim/neovim/issues/14090#issuecomment-913837694.

This fixes #55.
---
 lua/nvimpager.lua | 26 ++++++++++++++------------
 1 file changed, 14 insertions(+), 12 deletions(-)

diff --git a/lua/nvimpager.lua b/lua/nvimpager.lua
index 6737833..aa2c34a 100644
--- a/lua/nvimpager.lua
+++ b/lua/nvimpager.lua
@@ -390,18 +390,20 @@ end
 -- Detect possible filetypes for the current buffer by looking at the pstree
 -- or ansi escape sequences or manpage sequences in the current buffer.
 local function detect_filetype()
-  if not doc then
-    if detect_man_page_in_current_buffer() then
-      nvim.nvim_buf_set_option(0, 'filetype', 'man')
-    end
-  else
-    if doc == 'git' then
-      -- Use nvim's syntax highlighting for git buffers instead of git's
-      -- internal highlighting.
-      strip_ansi_escape_sequences_from_current_buffer()
-    elseif doc == 'pydoc' or doc == 'perldoc' or doc == 'ri' then
-      doc = 'man'
-    end
+  if not doc and detect_man_page_in_current_buffer() then doc = 'man' end
+  if doc == 'git' then
+    -- Use nvim's syntax highlighting for git buffers instead of git's
+    -- internal highlighting.
+    strip_ansi_escape_sequences_from_current_buffer()
+  end
+  if doc == 'man' then
+    nvim.nvim_buf_set_option(0, 'readonly', false)
+    nvim.nvim_command("Man!")
+    nvim.nvim_buf_set_option(0, 'readonly', true)
+  elseif doc == 'pydoc' or doc == 'perldoc' or doc == 'ri' then
+    doc = 'man' -- only set the syntax, not the full :Man plugin
+  end
+  if doc ~= nil then
     nvim.nvim_buf_set_option(0, 'filetype', doc)
   end
 end
